runtime/compare.c

```diff
@@ -99,6 +100,31 @@ static intnat compare_val(value v1, value v2, int total)
   return res;
 }
 
+static void poll_pending_actions(struct compare_stack* stk,
+                                 struct compare_item* next_stack_slot)
+{
+  if (caml_check_pending_actions()) {
+    value exn;
+    if (next_stack_slot == stk->stack) {
+      exn = caml_do_pending_actions_exn();
+    } else {
+      CAMLassert(next_stack_slot > stk->stack);
+      value* roots_start = (value*)(stk->stack);
+      size_t roots_length =
+        (next_stack_slot - stk->stack)
+        * sizeof(struct compare_item) / sizeof(value);
+      Begin_roots_block(roots_start, roots_length);
```

Comment:
+1 for using the old roots API, it works just fine here.  @damiendoligez is going to grumble about it, though.